vidarr-server/src/main/java/ca/on/oicr/gsi/vidarr/server/PriorityByWorkflow.java

```diff
@@ -122,22 +122,29 @@ public synchronized ConsumableResourceResponse request(
     }
 
     final var state = workflows.get(workflowName);
-    if (workflowPriority >= state.waiting.last().getValue()) {
-      state.waiting.remove(new SimpleEntry(vidarrId, workflowPriority));
-      currentInPriorityWaitingCount.labels(workflowName).set(state.waiting.size());
-      return ConsumableResourceResponse.AVAILABLE;
+    SimpleEntry resourcePair = new SimpleEntry(vidarrId, workflowPriority);
+
+    if (!state.waiting.isEmpty()) {
+      if (workflowPriority >= state.waiting.last().getValue()) {
+        state.waiting.remove(resourcePair);
+        currentInPriorityWaitingCount.labels(workflowName).set(state.waiting.size());
+        return ConsumableResourceResponse.AVAILABLE;
+      } else {
+        state.waiting.add(resourcePair);
+        currentInPriorityWaitingCount.labels(workflowName).set(state.waiting.size());
+        return ConsumableResourceResponse.error(
+            String.format("There are %s workflows currently queued up with higher priority.",
+                workflowName));
+      }
     } else {
-      state.waiting.add(new SimpleEntry(vidarrId, workflowPriority));
-      currentInPriorityWaitingCount.labels(workflowName).set(state.waiting.size());
-      return ConsumableResourceResponse.error(
-          String.format("There are %s workflows currently queued up with higher priority.", workflowName));
+        return ConsumableResourceResponse.AVAILABLE;
     }
```

Comment:
What do you think about this in place of lines 124-141? It does the checks for "is this the best workflow run to launch" first, and only adds to the waiting set if the answer is "no" (rather than adding to the waiting set first, then checking if it's the best workflow run to launch, and removing if yes)
```
    final var state = workflows.get(workflowName);

    if (state == null || state.waiting.isEmpty()) {
      return ConsumableResourceResponse.AVAILABLE;
    }
    if (workflowPriority >= state.waiting.last().getValue()) {
      return ConsumableResourceResponse.AVAILABLE;
    } else {
      set(workflowName, vidarrId, input);
      return ConsumableResourceResponse.error(
          String.format("There are %s workflows currently queued up with higher priority.",
              workflowName));
    }
```